### ISSUE TITLE
Error when running tests defined in a trait

### ISSUE THREAD

**Issue by ramondelafuente** [phpspec/phpspec#825]
`PHPSPec 2.4.0`
`PHP 5.7`

Sometimes when trying to tidy up specs we divide them into traits.

So for a class called `subject`

```
class Subject {
    public function something(\DateTime $d) {
        return true;
    }
    public function somethingElse(\DateTime $d) {
                return true;
    }
}
```

the specs will be in

```
class SubjectSpec {
    use SubjectElseTrait;

    it_does_something(\DateTime $d) {
        $this->something($d)->willReturn(true);
    }
}

trait SubjectElse {
    it_does_somethingElse(\DateTime $d) {
                $this->somethingElse($d)->willReturn(true);
    }
}
```

This is just to make smaller spec files (we have large blocks of XML defined as results of API calls so the spec files tend to grow large). This works fine with `PHPSPec 2.4.0` and `PHP 5.6`.

However, when we try to run these tests in `PHP7`, the arguments specified in traits are not "unpacked".

`Uncaught TypeError: Argument 1 passed to SubjectSpec::it_does_somethingElse() must be an instance of DateTime, instance of PhpSpec\Wrapper\Collaborator given`.

Is this a known issue?

**Comment by sagikazarmark** [comment]
This happened to me as well: https://travis-ci.org/php-http/utils/jobs/97324669

**Comment by sagikazarmark** [comment]
A workaround:

``` php
class SomeSpec
{
    use SomeBehavior {
        SomeBehavior::it_does_something as SoingSomethingExample;
    }

    /**
     * @see https://github.com/phpspec/phpspec/issues/825
     */
    function it_does_something(Something $something)
    {
        $this->SoingSomethingExample($stream);
    }
}
```

**Comment by sagikazarmark** [comment]
No, it's actually not working, but I don't get why.

**Comment by sagikazarmark** [comment]
Hm, it worked with one addition: you must remove the typehint from the method in the trait. You can add it back as soon as this is resolved.

**Comment by ciaranmcnulty** [comment]
This is because the way we fixed typehints for PHP7 doesn't take Traits into account

**Comment by ciaranmcnulty** [comment]
@sagikazarmark How is the trait being found in your example (spec folder is not autoloadable)?

**Comment by sagikazarmark** [comment]
Sorry, forgot to add that detail: I added the spec folder to dev autoloading in composer config.,

**Comment by ciaranmcnulty** [comment]
The error is that type hints don't work in anything autoloaded; this is not supported behaviour.

There is a workaround of manually including the trait in the following way:

```
use PhpSpec\Loader\StreamWrapper;
require_once StreamWrapper::wrapPath(__DIR__ . '/Path/To/MyTrait.php');
```

**Comment by stof** [comment]
@ciaranmcnulty what you could do is register an autoloader for the spec folder which applies this wrapper. This way, traits would be rewritten

**Comment by ciaranmcnulty** [comment]
Possibly but they are not auto loaded currently

**Comment by sagikazarmark** [comment]
Hm, is this resolved?

https://github.com/moneyphp/money/pull/220

**Comment by sagikazarmark** [comment]
Anyone?

**Comment by ciaranmcnulty** [comment]
Issue is still open - defining specs in traits isn't something we're supporting yet so it's 'undefined behaviour'.

**Comment by sagikazarmark** [comment]
Thanks for your response. If you take a look at the linked PR, it seems to work.

**Comment by ciaranmcnulty** [comment]
The errors are around the typehints on PHP7... if you're not using Typehints for doubles, or not using PHP7, you won't have an issue.

**Comment by sagikazarmark** [comment]
Ah, thanks.

**Comment by walterdolce** [comment]
Hi, just to add to the conversation as I found myself implementing (exctracting) shared object related verifications in traits - not in PHP 7.

It seems it is possible to avoid the nasty `require_once` suggested above by just adding the spec folder to the autoloader paths array in the `composer.json` file.

Example:
```
# composer.json
{
  "autoload": {
    "psr-0": {
      "": [
        "webapp/php/lib",
        "spec"
      ]
    },
}

===================

# project dir structure
webapp/
   php/
      lib/
         VendorNamespace/
            Html/
               Element/
                  Option.php
                  Select.php
                  Input.php
spec/
   VendorNamespace/
            Html/
               Element/
                  Traits/
                     ElementIdentityTrait.php   <--- shared object related verifications here
                  OptionSpec.php
                  SelectSpec.php
                  InputSpec.php
```

**Comment by ciaranmcnulty** [comment]
@walterdolce Yes this works fine if you're not using PHP 7

**Comment by ciaranmcnulty** [comment]
You should probably use `autoload-dev` for this BTW

**Comment by walterdolce** [comment]
> You should probably use autoload-dev for this BTW
Ah good point, thanks. 
Example above updated so future comers will know what/how to do (for < PHP 7).

> Yes this works fine if you're not using PHP 7
Oh I see. So the very same setup doesn't work in PHP 7?
Haven't tried it myself yet..

**Comment by ciaranmcnulty** [comment]
No, the problem is that in PHP7 we have to remove the type hints from the specs, if the class is loaded outside of PhpSpec (by include or by an autoloader) we can't process it.
